### DISCUSSION TITLE
Specifying dlls to be packed in the .net 6 csproj file

### DISCUSSION THREAD

**Discussion by JonTheTurnip** [Excel-DNA/ExcelDna#595] [Q&A] [answered]
Hi there,

I'm trying to transition my ExcelDna project from .net framework to .net 6.0, and in the process define my exceldna properties in the .csproj file rather than the .dna file.

I've removed the .dna file completely from the project and I've successfully added some dna properties as per the screenshot below.  However I would like to add the list of assemblies that need to be packed, if I don't do this then it seems that some assemblies get missed.

Please can someone advise on how to manually list the dlls to be packed?  As an example I tried using the ExcelAddInInclude property for a single dll that's in the bin folder, but I get an error.

Thanks! 

![image](https://user-images.githubusercontent.com/10959562/233040224-073fbf2e-14af-4193-a13b-37759e26dd6f.png)

![image](https://user-images.githubusercontent.com/10959562/233041335-d88e7dcf-6d69-4610-88ea-d6a68c1e88a6.png)

**Comment by govert** [comment, marked as answer]
TLDR; (1) You should not have the extra lines in the `ExcelAddInInclude` property. Make it like this:
```
<ExcelAddInInclude>Microsoft.Extensions.Logging.dll<ExcelAddInInclude>
```

TLDR; (2) But it should all work without the `ExcelAddInInclude`, if you are using the latest ExcelDna.AddIn package version 1.7.0-rc3.

It did take me a bit to notice this, so here is some of the missing documentation I wrote along the way.

---

If you are targeting .NET 6, then the assemblies (and also native runtime dependencies) to be packed should be picked up from the .deps.json file (be sure to use that latest 1.7.0-rc2 version to test this). If this does not work right for you, please give some more details for us to look at.

This dependency mapping from the .deps.json file can be disabled with these properties
```
    <ExcelDnaPackManagedDependencies>false</ExcelDnaPackManagedDependencies>
    <ExcelDnaPackNativeLibraryDependencies>false</ExcelDnaPackNativeLibraryDependencies>
```

You can add assemblies to the packing list with a property
```
    <ExcelAddInInclude>MyLibrary1.dll;MyLibrary2.dll;OtherDependency.dll</ExcelAddInInclude>
```

If you want to generate the packing list from the list of .dll files in the output, you can do this with a Target the builds the `ExcelAddInInclude` list dynamically
```
  <Target Name="PackedReferences" AfterTargets="AfterBuild" BeforeTargets="ExcelDnaBuild">
    <ItemGroup>
      <References Include="$(OutDir)*.dll" Exclude="$(OutDir)$(TargetFileName)"/>
    </ItemGroup>

    <PropertyGroup>
      <ExcelAddInInclude>@(References)</ExcelAddInInclude>
    </PropertyGroup>
  </Target>
```

The error you show is because you are explicitly adding Microsoft.Logging.Extensions.dll to the packing list, but it appears that this library is not found in the output directory. I think the reason has to do with the formatting of your property, leading to a the task looking for a file with spaces before and after the name. Hence the packing task is failing.

---

@Sergey-Vlasov Can we please trim the `ExcelAddInInclude` entries after splitting, so that the list can be formatted a bit more easily as
```
<ExcelAddInInclude>
    MyLibrary1.dll
<ExcelAddInInclude>
```
or
```
<ExcelAddInInclude>
    MyLibrary1.dll;
    MyLibrary2.dll
<ExcelAddInInclude>
```

Should we consider also supporting a line-end delimited list like this (without the semi-colons)?
```
<ExcelAddInInclude>
    MyLibrary1.dll
    MyLibrary2.dll
    OtherLibrary.dll
<ExcelAddInInclude>
```
Are there msbuild file conventions arounds this?

**Reply by JonTheTurnip** [reply]
Hi @govert , 

As per your second suggestion, I have just switched over to 1.7.0-rc3 and the packed xll seems to now work fine without having to manually list all references.   

There's just one issue that I don't understand.   When I run the unpacked xll I do not get the "A non collectible assembly may not reference a collectible assembly" error, however I do get that error on the packed xll  Any thoughts on why this might be?  Happy to provide more detail if needed.

**Reply by govert** [reply]
That error is exactly why the `ExcelAddInDisableAssemblyContextUnload` flag was added. See https://github.com/Excel-DNA/ExcelDna/issues/450 and https://github.com/Excel-DNA/ExcelDna/pull/462.
If you still have your own .dna file in the project, you need to manually add the `<DnaLibrary DisableAssemblyContextUnload="true" ...  />` attribute there. Else if you have no .dna file anymore, the property in the .csproj file should write this into the output .dna files.

It might further be that you need to make some calls (e.g. code that make make and XmlSerializer, or other dynamics types or loading) inside an `EnterContextualReflection` scope - see https://github.com/dotnet/runtime/blob/main/docs/design/features/AssemblyLoadContext.ContextualReflection.md

The problem is mostly that some code is generated and loaded into the default AssemblyLoadContext, instead of the (normally collectible) add-in ALC. But there might be other possibilities with code or assemblies considered 'not collectible'. Unloadable ALCs under .NET 6+ does not seem to have great support.

If you're stuck with this, maybe you can try to make a small reproduction and I'll have a look.
You're likely to at least pinpoint the dependency and call combination causing this.

**Reply by JonTheTurnip** [reply]
I did get rid of the .dna file entirely and I am specifying \<ExcelAddInDisableAssemblyContextUnload\>true\</ExcelAddInDisableAssemblyContextUnload\> in the csproj file.  Doing this does solve the issue for the unpacked xll but unfortunately not for the packed one.  

I don't understand why the difference in behavior between the packed/unpacked, but to be honest it's way beyond my knowledge of .net.  I did try and follow the .net doc that you linked to!

It's probably best that I stick to .net framework for now as I really do need the packed xll.

Thanks for your help, very much appreciated :)

**Reply by govert** [reply]
Could you possible make some reproduction of this that we can try? Or give a bit more information about your dependencies and anything 'dynamic' that you think might be happening?
Sticking with .NET Framework for now is not bad, but it would be nice if we can improve the .NET 6 support for the future.

**Reply by govert** [reply]
OK - I can see a bug with this in the packed add-in. We'll fix that for the next pre-release version.

**Reply by JonTheTurnip** [reply]
Oh, brilliant!  Do you know when that will be?  I'll make sure to grab the latest, test it out and let you know if it fixes my issue.

**Reply by JonTheTurnip** [reply]
@govert , just wanted to let you know that I saw that there was a new package, '-rc4', available.  I've updated my project to use the new package and I can confirm that the "A non collectible assembly may not reference a collectible assembly" error no longer occurs.  Many thanks for fixing this :)

**Comment by govert** [comment]
@Sergey-Vlasov Looking at https://learn.microsoft.com/en-us/visualstudio/msbuild/comparing-properties-and-items?source=recommendations&view=vs-2022 I think we don't have to support the list without the semi-colons - the semi-colons seem standard, but they do have an example with an extra semi-colon at the end,
```
<PropertyGroup>
    <BuildDependsOn>
        $(BuildDependsOn);
        CustomBuild;
    </BuildDependsOn>
</PropertyGroup>
```
so we should support something like
```
<ExcelAddInInclude>
    MyLibrary1.dll;
    MyLibrary2.dll;
<ExcelAddInInclude>
```

**Comment by Turhan8** [comment]
how can I pack the appsettings.json into the xll?
we have ExternalLibrary, Reference and Image tags, but is there anything for text content to be packed?

**Comment by govert** [comment]
I suggest that additional files like that should be built into your own .dll file as resources. This uses standard .NET mechanism as described here:
https://learn.microsoft.com/en-us/dotnet/core/extensions/resources

**Reply by Turhan8** [reply]
Also it makes sense to have config file outside in case you want to modify it. I only asked because it used to pack the config files before, in .net framework version.
